### PR TITLE
test/bluetooth/tester: Re-pair on lost bond

### PR DESCRIPTION
**Pull request by sjanc**
If IUT is ascting as a central device and peer lost bond we need
to re-pair to restore bond. PTS is not sending any WID for this
so bondingn needs to be initiated implicitly by IUT.

This was affecting GAP/SEC/AUT/BV-25-C qualification test case.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

### CODE REVIEW

**Review by jhedberg — APPROVED** [review]

File: tests/bluetooth/tester/src/gap.c [hunk 1]
```diff
@@ -171,11 +172,24 @@ static void le_security_changed(struct bt_conn *conn, bt_security_t level,
 			    CONTROLLER_INDEX, (uint8_t *) &sec_ev, sizeof(sec_ev));
 		break;
 	case BT_SECURITY_ERR_PIN_OR_KEY_MISSING:
-		memcpy(bond_ev.address, addr->a.val, sizeof(bond_ev.address));
-		bond_ev.address_type = addr->type;
+		bt_conn_get_info(conn, &info);
```
Comment (Thalley) [inline tests/bluetooth/tester/src/gap.c:175]: Please either store the result of this in a variable and potentially log if it returns an error, or add `(void)` to explicitly discard the return value. 

This comment also applies to `memcpy` in line 185, and `bt_conn_set_security` in line 191

File: tests/bluetooth/tester/src/gap.c [hunk 2]
```diff
@@ -171,11 +172,24 @@ static void le_security_changed(struct bt_conn *conn, bt_security_t level,
...
+		if (info.role == BT_CONN_ROLE_CENTRAL) {
+			LOG_DBG("Bond lost");
 
-		tester_send(BTP_SERVICE_ID_GAP, GAP_EV_BOND_LOST,
-			    CONTROLLER_INDEX, (uint8_t *) &bond_ev, sizeof(bond_ev));
+			memcpy(bond_ev.address, addr->a.val, sizeof(bond_ev.address));
+			bond_ev.address_type = addr->type;
+
+			tester_send(BTP_SERVICE_ID_GAP, GAP_EV_BOND_LOST,
+				    CONTROLLER_INDEX, (uint8_t *) &bond_ev, sizeof(bond_ev));
```
Comment (Thalley) [inline tests/bluetooth/tester/src/gap.c:189]:
```suggestion
				    CONTROLLER_INDEX, (uint8_t *)&bond_ev, sizeof(bond_ev));
```

File: tests/bluetooth/tester/src/gap.c [hunk 3]
```diff
@@ -159,6 +159,7 @@ static void le_security_changed(struct bt_conn *conn, bt_security_t level,
 	const bt_addr_le_t *addr = bt_conn_get_dst(conn);
 	struct gap_sec_level_changed_ev sec_ev;
 	struct gap_bond_lost_ev bond_ev;
+	struct bt_conn_info info;
```
Comment (Thalley) [inline tests/bluetooth/tester/src/gap.c:162]: This is only used within the `case BT_SECURITY_ERR_PIN_OR_KEY_MISSING:` Consider moving the declaration to there.

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bluetooth/tester/src/gap.c (hunk 1 above)
Comment (sjanc) [reply tests/bluetooth/tester/src/gap.c]: memcpy doesn't return any error

File: tests/bluetooth/tester/src/gap.c (hunk 3 above)
Comment (sjanc) [reply tests/bluetooth/tester/src/gap.c:162]: this will not compile (unless extra { } are used)

File: tests/bluetooth/tester/src/gap.c (hunk 1 above)
Comment (Thalley) [reply tests/bluetooth/tester/src/gap.c]: It doesn't return an error, but it does return a pointer: 
```
void *memcpy(void *_MLIBC_RESTRICT d, const void *_MLIBC_RESTRICT s, size_t n)
```

Regardless of the return type, we should add `(void)` if we explicitly ignore that value.

File: tests/bluetooth/tester/src/gap.c (hunk 3 above)
Comment (Thalley) [reply tests/bluetooth/tester/src/gap.c:162]: I know :) 

It's fine to keep as is

File: tests/bluetooth/tester/src/gap.c (hunk 1 above)
Comment (sjanc) [reply tests/bluetooth/tester/src/gap.c]: fair enough (although it is not very common in zephyr code)

Comment (Thalley) [reply tests/bluetooth/tester/src/gap.c]: It is very common for e.g. `memset`, but for some reason it hasn't been done that much for e.g. `memcpy`.

File: tests/bluetooth/tester/src/gap.c [hunk 4]
```diff
@@ -171,11 +172,23 @@ static void le_security_changed(struct bt_conn *conn, bt_security_t level,
...
+
+			(void)memcpy(bond_ev.address, addr->a.val, sizeof(bond_ev.address));
+			bond_ev.address_type = addr->type;
 
-		tester_send(BTP_SERVICE_ID_GAP, GAP_EV_BOND_LOST,
-			    CONTROLLER_INDEX, (uint8_t *) &bond_ev, sizeof(bond_ev));
+			tester_send(BTP_SERVICE_ID_GAP, GAP_EV_BOND_LOST,
+				    CONTROLLER_INDEX, (uint8_t *)&bond_ev, sizeof(bond_ev));
+
+			bt_conn_set_security(conn, BT_SECURITY_L2 | BT_SECURITY_FORCE_PAIR);
```
Comment (Thalley) [inline tests/bluetooth/tester/src/gap.c:190]:
```suggestion
			(void)bt_conn_set_security(conn, BT_SECURITY_L2 | BT_SECURITY_FORCE_PAIR);
```
or store the result and log error if any.

**Review by Thalley — CHANGES_REQUESTED** [review]

**Review by Thalley — APPROVED** [review]

**Pull request merged**